src/main/java/io/stargate/sgv2/jsonapi/service/processor/CommandProcessor.java

```diff
@@ -46,36 +60,92 @@ public CommandProcessor(
 
-        // if we have a non-null item
-        // call supplier get to map to the command result
-        .onItem()
-        .ifNotNull()
-        .transform(Supplier::get);
+                        } else {
+                          // return the tags and command result
+                          return item;
+                        }
+                      })
+                  .onItem()
+                  .ifNotNull()
+                  .transform(Supplier::get)
+                  .onItem()
+                  .transform(
+                      result -> {
+                        Tag errorTag = MetricsConstants.ERROR_FALSE_TAG;
+                        ;
+                        Tag errorClassTag = MetricsConstants.DEFAULT_ERROR_CLASS_TAG;
+                        Tag errorCodeTag = MetricsConstants.DEFAULT_ERROR_CODE_TAG;
+                        if (null != result.errors() && !result.errors().isEmpty()) {
+                          errorTag = MetricsConstants.ERROR_TRUE_TAG;
+                          String errorClass =
+                              (String)
+                                  result
+                                      .errors()
+                                      .get(0)
+                                      .fields()
+                                      .getOrDefault(
+                                          MetricsConstants.ERROR_CLASS,
+                                          MetricsConstants.UNKNOWN_VALUE);
+                          errorClassTag = Tag.of(MetricsConstants.ERROR_CLASS, errorClass);
+                          String errorCode =
+                              (String)
+                                  result
+                                      .errors()
+                                      .get(0)
+                                      .fields()
+                                      .getOrDefault(
+                                          MetricsConstants.ERROR_CODE,
+                                          MetricsConstants.UNKNOWN_VALUE);
+                          errorCodeTag = Tag.of(MetricsConstants.ERROR_CODE, errorCode);
+                        }
+                        Tags tags =
+                            Tags.of(commandTag, tenantTag, errorTag, errorClassTag, errorCodeTag);
+                        // add metrics
+                        meterRegistry
+                            .counter(MetricsConstants.COUNT_METRICS_NAME, tags)
+                            .increment();
+                        meterRegistry
+                            .timer(MetricsConstants.TIMER_METRICS_NAME, tags)
+                            .record(
+                                System.currentTimeMillis() - start,
+                                java.util.concurrent.TimeUnit.MILLISECONDS);
+                        // return the command result
+                        return result;
+                      });
+            });
```

Comment:
First of all I still think metrics could be added in the `MeteredCommandProcessor`, which wraps the current one and does the metrics only.. But even if you don't want to do this, the code should be simply something like:

```
public <T extends Command> Uni<CommandResult> processCommand(
      CommandContext commandContext, T command) {

Timer.Sample sample = Timer.start(registry);

// same code as now
return ...

         // if we have a non-null item
        // call supplier get to map to the command result
        .onItem()
        .ifNotNull()
        .transform(Supplier::get)

        .onItem()
        .call(commandResult -> {
            Tags tags = resolveTags(command, commandResult);
            sample.stop(registry.timer("my.timer", tags));
        });
```

Resolving tags should be extracted to a method for clarity and easier reading.

Reply:
Again having `MeteredCommandProcessor` seems like a better solution, as you can wrap the current one and thus extract only metric catching to this class.. This also enables you to easily tests everything cause you can always mock the delegating CommandProcessor and thus flush different results..

```
public <T extends Command> Uni<CommandResult> processCommand(
      CommandContext commandContext, T command) {

Timer.Sample sample = Timer.start(registry);

return delegate.processCommand(commandContext, command)

        .onItem()
        .call(commandResult -> {
            Tags tags = resolveTags(command, commandResult);
            sample.stop(registry.timer("my.timer", tags));
        });
```